pkg/querier/store_gateway_client.go

```diff
@@ -100,12 +100,21 @@ func newStoreGatewayClientPool(discovery client.PoolServiceDiscovery, clientConf
 	return client.NewPool("store-gateway", poolCfg, discovery, newStoreGatewayClientFactory(clientCfg, reg), clientsCount, logger)
 }
 
+// PoolConfig is config for creating a Pool of gRPC clients.
+type PoolConfig struct {
+	CleanupPeriod time.Duration `yaml:"cleanup_period" category:"advanced"`
+	RemoteTimeout time.Duration `yaml:"remote_timeout" category:"advanced"`
+}
+
 type ClientConfig struct {
 	TLSEnabled bool             `yaml:"tls_enabled" category:"advanced"`
 	TLS        tls.ClientConfig `yaml:",inline"`
+	PoolConfig PoolConfig       `yaml:",inline"`
 }
 
 func (cfg *ClientConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.BoolVar(&cfg.TLSEnabled, prefix+".tls-enabled", cfg.TLSEnabled, "Enable TLS for gRPC client connecting to store-gateway.")
+	f.DurationVar(&cfg.PoolConfig.CleanupPeriod, prefix+".cleanup-period", 5*time.Second, "How frequently to clean up clients for store-gateways that have gone away.")
+	f.DurationVar(&cfg.PoolConfig.RemoteTimeout, prefix+".remote-timeout", time.Second, "Timeout for detecting a store-gateway that has gone away.")
```

Comment:
This seems very low. I would not merge a default of 1s. Let's keep the current 10s, but test this in grafana.com with explicit setting to 1s.
I'd be afraid of breaking things for users when they have high loads.

Reply:
it's already 1s on the querier, but I'm ok to first merge the config options and then change the defaults

Reply:
wait. if we test in grafana cloud with timeout of 10s, this still doesn't protect any user who runs this on prem. And the timeout for the querier->query-frontend is already the same.

@pstibrany do you have opinions on the length of the timeout?

> > We can use `DialContext` with a context with an expiration and also use `WithBlock()` here:
> 
> This sounds like a good idea, but will require some changes to the client pool. We already get similar benefit from using time-bound context for individual calls (if connection doesn't exist yet, it is created first during call to eg. `Series` or other methods on `StoreGatewayClient` interface).

Reply:
> it's already 1s on the querier, but I'm ok to first merge the config options and then change the defaults

I only see 10s in the querier: https://github.com/grafana/mimir/blob/86dc0f073250b59b674592d3d43e9c1781806d88/pkg/querier/store_gateway_client.go#L90 

Maybe I'm looking at the wrong place?

Reply:
you're looking in [store_gateway_client.go](https://github.com/grafana/mimir/blob/86dc0f073250b59b674592d3d43e9c1781806d88/pkg/querier/store_gateway_client.go#L90) which is also what this PR is changing. I was referring to this piece https://github.com/grafana/mimir/blob/006f62659fda18c77fbf77c53e504dab9044e7db/pkg/querier/worker/scheduler_processor.go#L64-L68